### PR TITLE
table designer perf improvement - add user setting for preload db model 

### PR DESCRIPTION
**Pull request by alanrenmsft**
Currently, the initial launch of table designer takes more time than subsequent ones, introduce a user setting to let user decide whether preload of the database model should be performed to make the initial launch experience smoother.

when user opens table designer (for both new and existing tables) for the first time, we show the following prompt.
image

regardless of the choice, this prompt will not show again.

STS changes since current version:
image

### CODE REVIEW

**Comment by Charles-Gagnon** [comment]
So what happens if the user opens the node, then walks away for a bit and meanwhile changes are made to the table on the server? Will we have stale data at that point? (and if not, what exactly is this preloading doing then if it still has to get the "latest" schema when you start the edit)

**Comment by alanrenmsft** [comment]
> So what happens if the user opens the node, then walks away for a bit and meanwhile changes are made to the table on the server? Will we have stale data at that point? (and if not, what exactly is this preloading doing then if it still has to get the "latest" schema when you start the edit)

@Charles-Gagnon, when we launch table designer, we will check whether the table designer related schema information is still up to date by running a query that returns a string (https://msdata.visualstudio.com/SQLToolsAndLibraries/_git/DacFx?path=/Source/DesignServices/TableDesignerService/TableDesignerCacheManager.cs&version=GBmain&line=31&lineEnd=32&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents), if the version string matches the current model then proceed otherwise reload the database model to make sure user is working with the newest schema. 

I don't know in real world, how frequently users will run into this scenario, maybe we can add more description to the setting so that users can better decide whether they want to turn this on or not?

**Review by caohai — APPROVED** [review]

**Comment by Charles-Gagnon** [comment]
Yeah, hopefully it won't happen all that often. I'm slightly worried about perf impact if someone opens up a database with hundreds or thousands of tables though so maybe it'd be good to call out potential perf impacts. What about letting users choose what tables are preloaded? So like when they open a table for design it'll ask if they want to preload it in the future - and then we just keep a list of all the tables to preload on expansion. You could still have an option to let them select "all" and then just give a big ole warning about potential perf issues.

**Comment by caohai** [comment]
> Yeah, hopefully it won't happen all that often. I'm slightly worried about perf impact if someone opens up a database with hundreds or thousands of tables though so maybe it'd be good to call out potential perf impacts.
> 
> Or what about letting users choose what tables are preloaded? So like when they open a table for design it'll ask if they want to preload it in the future - and then we just keep a list of all the tables to preload on expansion. You could still have an option to let them select "all" and then just give a big ole warning about potential perf issues.

I've actually done some basic benchmark for loading db with >1000 tables. The load time (around 10s) is not significantly higher than loading a db with say 50 tables. I think the underlying reason is that DacFx is sending queries to db in batch.

File: extensions/mssql/package.nls.json [hunk 1]
```diff
@@ -188,5 +188,6 @@
 
 	"title.newTable": "New Table",
 	"title.designTable": "Design",
-	"mssql.parallelMessageProcessing" : "[Experimental] Whether the requests to the SQL Tools Service should be handled in parallel. This is introduced to discover the issues there might be when handling all requests in parallel. The default value is false. Relaunch of ADS is required when the value is changed."
+	"mssql.parallelMessageProcessing" : "[Experimental] Whether the requests to the SQL Tools Service should be handled in parallel. This is introduced to discover the issues there might be when handling all requests in parallel. The default value is false. Relaunch of ADS is required when the value is changed.",
+	"mssql.tableDesigner.preloadDatabaseModel": "Whether to preload the database model when the database node in the object explorer is expanded. When enabled, the loading time of table designer can be reduced."
```
Comment (Charles-Gagnon) [inline extensions/mssql/package.nls.json:192]: Suggest adding warning about potential perf (CPU, memory) hits by doing this.

**Review by Charles-Gagnon — APPROVED** [review]
Sounds good, thanks for investigating that @caohai. I still think that it wouldn't hurt to add that feature but maybe wait and see if people have any noticeable perf issues after enabling this before working on such a thing.

**Review by kisantia — APPROVED** [review]

**Comment by Charles-Gagnon** [comment]
Ohhhhh, yeah ok I was not aware of that. Gotcha. And to make sure - this only happens when you expand the Database itself - not when you expand the Databases folder node? That should be fine then, we're already querying a bunch of stuff when that happens so a bit more probably won't hurt much.

**Comment by alanrenmsft** [comment]
@Charles-Gagnon yes, only for database node: https://github.com/microsoft/sqltoolsservice/pull/1690/files#diff-ea0c018613c1eb7df2b5cc4259691e9477700a853fdec503c3b8593ea96d22a0R387

**Pull request merged**